tests/MLSearch.UnitTests/Indexing/Initializer/InfiniteChatSequenceTests.cs

```diff
@@ -0,0 +1,313 @@
+using ActualChat.Chat;
+using ActualChat.MLSearch.Indexing.Initializer;
+using System.Linq.Expressions;
+
+namespace ActualChat.MLSearch.UnitTests.Indexing.Initializer;
+
+internal static class SequenceHelpers
+{
+    public static async Task EnumerateAll<T>(this IAsyncEnumerable<T> sequence)
+    {
+        await foreach (var _ in sequence) { }
+    }
+}
+
+public class InfiniteChatSequenceTests
+{
+    private static readonly IMomentClock Clock = Mock.Of<IMomentClock>();
+    private static readonly IChatsBackend Chats = Mock.Of<IChatsBackend>();
+    private static readonly ILogger<InfiniteChatSequence> Log = Mock.Of<ILogger<InfiniteChatSequence>>();
+    private static readonly Expression<Func<IChatsBackend, Task<ApiArray<Chat.Chat>>>> ListChangedCall =
+        x => x.ListChanged(
+            It.IsAny<long>(),
+            It.IsAny<long>(),
+            It.IsAny<ChatId>(),
+            It.IsAny<int>(),
+            It.IsAny<CancellationToken>());
+
+    [Fact]
+    public async Task ChatsAreBeingLoadedInBatches()
+    {
+        var chats = new Mock<IChatsBackend>();
+        chats
+            .Setup(ListChangedCall)
+            .Returns<long, long, ChatId, int, CancellationToken>(
+                (minVersion, _, _, size, _) => GetNextBatch(minVersion, size)
+            );
+
+        const int batchSize = 5;
+
+        var sequence = new InfiniteChatSequence(Clock, chats.Object, Log) {
+            BatchSize = batchSize,
+        };
+
+        const int maxChatCount = 23;
+        var prevVersion = 0L;
+        await foreach (var (_, version) in sequence.LoadAsync(1, CancellationToken.None).Take(maxChatCount)) {
+            Assert.True(version > prevVersion);
+            prevVersion = version;
+        }
+
+        chats.Verify(
+            ListChangedCall,
+            Times.Exactly((maxChatCount + batchSize - 1) / batchSize)
+        );
+    }
+
+    [Fact]
+    public async Task ThereIsADelayBeforeRetryingInCaseOfAnEmptyBatch()
+    {
+        var emptyBatchDelay = TimeSpan.FromSeconds(11111);
+
+        var clock = new Mock<IMomentClock>();
+        clock
+            .Setup(x => x.Delay(It.IsAny<TimeSpan>(), It.IsAny<CancellationToken>()))
```

Comment:
afaik in such case we write the first method call on the same line as variable

Reply:
It looks uglier in this certain case. I decided to keep this style for now.